pcdsdevices/signal.py

```diff
@@ -157,55 +194,222 @@ def _calc_readback(self):
+    subscribe.__doc__ = ophyd.ophydobj.OphydObject.subscribe.__doc__
+
+    def _setup_subscriptions(self):
+        """Subscribe to all relevant signals."""
+        if self._has_subscribed:
+            return
+
+        self._has_subscribed = True
+
+        try:
+            for signal, siginfo in self._signals.items():
+                self.log.debug("Subscribing %s", signal.name)
+                if siginfo.value_cbid is not None:
+                    # Only subscribe once.
+                    continue
+
+                siginfo.meta_cbid = signal.subscribe(
+                    self._signal_meta_callback,
+                    run=True,
+                    event_type=signal.SUB_META,
+                )
+
+                # Run the subscription now, only if it's a signal that doesn't
+                # normally run callbacks.
+                siginfo.value_cbid = signal.subscribe(
+                    self._signal_value_callback,
+                    run=is_meek_ophyd_object(signal),
+                )
+        except Exception:
+            self.log.exception("Failed to subscribe to signals")
+
+    def wait_for_connection(self, *args, **kwargs):
+        """Wait for underlying signals to connect."""
+        self._setup_subscriptions()
+        return super().wait_for_connection(*args, **kwargs)
+
+    def _signal_meta_callback(
+        self, *, connected: bool = False, obj: Signal, **kwargs
+    ) -> None:
+        """This is a SUB_META callback from one of the aggregated signals."""
+        with self._check_connectivity():
+            self._signals[obj].connected = connected
+
+    def _signal_value_callback(self, *, obj: Signal, **kwargs):
+        """This is a SUB_VALUE callback from one of the aggregated signals."""
         kwargs.pop('sub_type')
-        sig = kwargs.pop('obj')
         kwargs.pop('old_value')
         value = kwargs['value']
         with self._lock:
             old_value = self._readback
             # Update just one value and assume the rest are cached
             # This allows us to run subs without EPICS gets
-            value = self._insert_value(sig, value)
+            # Run metadata callbacks before the value callback, if appropriate
+            with self._check_connectivity() as connectivity_info:
+                value = self._insert_value(obj, value)
+            if connectivity_info["sent_value_callback"]:
+                # Avoid sending a duplicate SUB_VALUE event since the
+                # connectivity check above did it already
+                return
             if value != old_value or not self._update_only_on_change:
                 self._run_subs(sub_type=self.SUB_VALUE, obj=self, value=value,
                                old_value=old_value)
 
+    @property
+    def connected(self) -> bool:
+        """Are all relevant signals connected?"""
+        # Subscriptions are used to see if the device is connected:
+        self._setup_subscriptions()
```

Comment:
I'm not sure if its appropriate to set up EPICS anything inside of a property due to network latency. Is this necessary/are my concerns unfounded?
Are our subscriptions set up on an as-needed basis?

Reply:
I agree, but this was due to being painted in a corner:

* `.wait_for_connection()` on a device says "ensure that all components are connected by checking `.connected`"
* A call to `.connected` on a `MultiDerivedSignal` as-implemented returns the connectivity status based on whether or not subscriptions have been configured
* Without subscriptions configured, `.connected` will always return `False` and as such `wait_for_connection()` will raise a `TimeoutError`

Alternatively, perhaps, an entirely separate code path could be added for being used solely with `.get()` and `.put()` and never using subscriptions. I feel that this will likely make things more complicated logic-wise, but it could be doable if we determine the concerns are strong enough for it.

Reply:
I took the alternative approach from my above comment in [7f6c62c](https://github.com/pcdshub/pcdsdevices/pull/970/commits/7f6c62c2c363e7fad836e47231c441372418e88d); what do you think?

Reply:
It's not clear to me if any of these possible approaches is better than the others.

I guess the gap in my head that prompted this question was the understanding that "normal" epics signals start initiating the connection process in a background thread after init, whereas this multi derived signal seemed to be doing it on first request instead (because it needs to do a bit more than a normal epics signal?), making stuff like this kind of wonky.